### PR TITLE
Session tracking on by default

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Enables automatic session tracking by default.

### Added

`setEndpoints` API on configuration class.

### Changed

- Auto capture enabled by default
- Deprecated `setEndpoint` and `setSessionEndpoint`

## Tests

Added and ran unit tests, as per the design spec.

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: src/main/java/com/bugsnag/Configuration.java [hunk 1]
```diff
@@ -97,6 +102,52 @@ public boolean shouldAutoCaptureSessions() {
...
+            }
+        }
+
+        boolean invalidSessionsEndpoint = sessions == null || "".equals(sessions);
+        String sessionEndpoint = null;
+
+        if (invalidSessionsEndpoint) {
+            LOGGER.warn("The session tracking endpoint has not been"
+                    + " set. Session tracking is disabled");
+            this.autoCaptureSessions.set(false);
```
Comment (Pezzah) [inline src/main/java/com/bugsnag/Configuration.java]: Should autoCaptureSessions also be disabled if setSessionEndpoint is called with empty / null?

File: src/main/java/com/bugsnag/Configuration.java [hunk 2]
```diff
@@ -97,6 +102,52 @@ public boolean shouldAutoCaptureSessions() {
...
+        } else {
+            sessionEndpoint = sessions;
+        }
+
+        if (sessionDelivery instanceof HttpDelivery) {
+            ((HttpDelivery) sessionDelivery).setEndpoint(sessionEndpoint);
```
Comment (martin308) [inline src/main/java/com/bugsnag/Configuration.java]: why does it set the endpoint here when it is possible that we have an invalid sessions endpoint? If we want to do this it might be worth adding a code comment as to why

File: src/main/java/com/bugsnag/Configuration.java [hunk 3]
```diff
@@ -97,6 +102,52 @@ public boolean shouldAutoCaptureSessions() {
...
+            throw new IllegalArgumentException("Notify endpoint cannot be empty or null.");
+        } else {
+            if (delivery instanceof HttpDelivery) {
+                ((HttpDelivery) delivery).setEndpoint(notify);
+            } else {
+                LOGGER.warn("Delivery is not instance of HttpDelivery, cannot set notify endpoint");
+            }
+        }
+
+        boolean invalidSessionsEndpoint = sessions == null || "".equals(sessions);
```
Comment (martin308) [inline src/main/java/com/bugsnag/Configuration.java]: is it better to use [`isEmpty()`](https://docs.oracle.com/javase/6/docs/api/java/lang/String.html#isEmpty())?

File: src/main/java/com/bugsnag/Configuration.java (hunk 1 above)
Comment (fractalwrench) [reply src/main/java/com/bugsnag/Configuration.java]: Resolved by in-person conversation - this is intended behaviour for the changeset

**Comment by fractalwrench** [comment]
Addressed inline comments and added an extra test for the case when `Delivery` was implemented rather than `HttpDelivery`.

File: src/main/java/com/bugsnag/Configuration.java [hunk 4]
```diff
@@ -97,6 +102,54 @@ public boolean shouldAutoCaptureSessions() {
...
+     * Note that if you are setting a custom {@link Delivery}, this method should be called after
+     * the custom implementation has been set.
+     *
+     * @param notify the notify endpoint
+     * @param sessions the sessions endpoint
+     *
+     * @throws IllegalArgumentException if the notify endpoint is empty or null
+     */
+    public void setEndpoints(String notify, String sessions) throws IllegalArgumentException {
+        if (notify == null || notify.length() == 0) {
```
Comment (martin308) [inline src/main/java/com/bugsnag/Configuration.java]: is there a reason that we shouldn't be using [`isEmpty()`](https://docs.oracle.com/javase/6/docs/api/java/lang/String.html#isEmpty())?

File: src/main/java/com/bugsnag/Configuration.java [hunk 5]
```diff
@@ -97,6 +102,54 @@ public boolean shouldAutoCaptureSessions() {
...
+            LOGGER.warn("The session tracking endpoint has not been"
+                    + " set. Session tracking is disabled");
+            this.autoCaptureSessions.set(false);
+        } else {
+            sessionEndpoint = sessions;
+        }
+
+        if (sessionDelivery instanceof HttpDelivery) {
+            // sessionEndpoint may be invalid (i.e. null) here, which should result in the default
+            // HttpDelivery throwing a MalformedUrlException which prevents delivery.
```
Comment (martin308) [inline src/main/java/com/bugsnag/Configuration.java]: are we sure this is what we want to be doing in this case? It seems less optimal to use invalid data and rely on exceptions being thrown to prevent delivery

File: src/test/java/com/bugsnag/ConfigurationTest.java [hunk 6]
```diff
@@ -40,4 +55,95 @@ public void testSessionApiHeaders() throws java.lang.Exception {
...
+            public void deliver(Serializer serializer, Object object, Map<String, String> headers) {
+            }
+
+            @Override
+            public void close() {
+            }
+        };
+        config.delivery = delivery;
+        config.sessionDelivery = delivery;
+        config.setEndpoints("http://example.com", "http://sessions.example.com");
```
Comment (martin308) [inline src/test/java/com/bugsnag/ConfigurationTest.java]: do we need an assertion here?

**Comment by fractalwrench** [comment]
@martin308 thanks for the comments. I've changed `SyncHttpDelivery` to fail faster if the endpoint is null, and read stderr to check that a message is logged when the session endpoint is invalid.
